src/ape/api/compiler.py

```diff
@@ -51,6 +51,45 @@ def compile(
             List[:class:`~ape.type.contract.ContractType`]
         """
 
+    @raises_not_implemented
+    def get_imports(
+        self, contract_filepaths: List[Path], base_path: Optional[Path]
+    ) -> Dict[str, List[str]]:
+        """
+        Returns a list of imports for each contract in a given compiler.
+
+        Args:
+            contract_filepaths (List[pathlib.Path]): A list of source file paths to compile.
+            base_path (Optional[pathlib.Path]): Optionally provide the base path, such as the
+              project ``contracts/`` directory. Defaults to ``None``. When using in a project
+              via ``ape compile``, gets set to the project's ``contracts/`` directory.
+
+        Returns:
+            Dict[str, List[str]]
+        """
+
+    def _get_filename_dict_from_paths(
+        self, paths: List[Path], base_path: Optional[Path]
+    ) -> Dict[str, List[str]]:
+        """
+        Structure for getting a list of paths related to a filename.
+        Used to create a broad/generous assumption of related paths for a filename.
+        """
+        filename_dict: Dict[str, List[str]] = {}
+
+        for p in paths:
+            filename = str(p).split("/")[-1]
+
+            if not filename_dict.get(filename):
+                filename_dict[filename] = []
```

Comment:
This should really be using the `contract_type.source_id` instead of the file name.
Some languages can have the same named file in multiple directories.

Reply:
Yeah, I agree about the multiple files' paths possibly having the same filename. The issue I was coming up against is, at this point, I'm reading import strings out of a file and I don't yet have an object. I was trying to take an approach where I might be re-compiling files that don't need it, but grabbing any files matching the filename. Do you have a different approach in mind?
This is a general comment to your other suggestions as well. I was hoping to use the logic of looking at filenames from import statements as a shared method for all compiler languages. Each plugin would implement the "cleaning" of the import string however it is needed, but ultimately to grab source paths related to a given filename and recompile them. I'm very open to suggestions :)

Reply:
Thanks for meeting and discussing this ticket. Here's a brief recap:
Don't use filenames
Convert import strings to resolved paths
Convert paths to source_ids
Use source_ids in imports and references dicts for both key and list value